Ports/rudix/rudix.1
.\"
.\" Copyright (c) 2005-2012 Ruda Moura
.\"
.Dd June 10, 2012
.Dt RUDIX 1
.Os
.Sh NAME
.Nm rudix
.Nd Rudix Package Manager
.Sh SYNOPSIS
.Nm
.Op Fl h | v | a | A | l | R | K | u | z
.Op Fl i Ar pkg-path
.Op Fl r Ar package-id
.Op Fl I Ar package-id
.Op Fl L Ar package-id
.Op Fl s Ar package-id
.Op Fl S Ar path
.Op Fl V Ar package-id
.Op Fl f Ar package-id
.Op Fl n Ar package-id
.Sh DESCRIPTION
The
.Nm
utility is the package manager for Rudix therefore you can use it to install, remove, print information about packages and search files under the Rudix domain. It is possible to verify the integrity of installed packages and to fix (repair) packages.
.Pp
Every package has a unique identifier called \fIpackage-id\fP and you may refer to Rudix packages by using its full \fIpackage-id\fP or with its short alias. For example, package Compiler Cache's \fIpackage-id\fP is \fIorg.rudix.pkg.ccache\fP but you can use \fIccache\fP as alias.
.Pp
Without any arguments provided,
.Nm
prints the current list of packages installed.
.Pp
The options are as follows:
.Bl -tag -width indent
.It Fl h
Print help on screen and exit.
.It Fl v
Print version and copyright information.
.It Fl a
Print all packages available for installation (name-version-release).
.It Fl A
Print all alternative aliases.
.It Fl l
Print a list of packages with \fIpackage-id\fP, version and install date.
.It Fl i Ar pkg-path
Install package. If \fIpkg-path\fP is not a file then try to install the package from the Internet. This operation requires root privileges.
.It Fl n Ar package-id
Remote install package. This operation requires root privileges.
.It Fl u
Remote update installed packages. This operation requires root privileges.
.It Fl r Ar package-id
Remove package. This operation requires root privileges.
.It Fl R
Remove all installed packages. This operation requires root privileges.
.It Fl I Ar package-id
Print \fIpackage-id\fP version and install date.
.It Fl L Ar package-id
List content of package (files).
.It Fl s Ar package-id
List available versions for \fIpackage-id\fP.
.It Fl S Ar path
Print all packages that contains path.
.It Fl V Ar package-id
Verify \fIpackage-id\fP.
.It Fl K
Verify all installed packages.
.It Fl f Ar package-id
Fix (repair) \fIpackage-id\fP. This operation requires root privileges.
.It Fl z
Enter interactive mode (type exit to quit).
.El
.Sh COMMANDS
The
.Nm
utility accepts simple commands which are related to the options seen above.
The list of commands are: help, version, available, aliases, list, remove-all, verify-all, update, interactive, info, files, install, remove, search, owner, verify, fix.
.Sh SEE ALSO
.Xr pkgutil 1 ,
.Xr installer 8 ,
.Xr repair_packages 8 ,
.Xr packagemaker 1


ChangeLog.txt
CHANGES FROM RUDIX 2009 TO RUDIX 2010

- Everything built on Snow Leopard.

CHANGES FROM RUDIX 2008r3 TO RUDIX 2009

- Everything built on Leopard.

CHANGES FROM RUDIX 2008r2 TO RUDIX 2008r3

* Build System

- Managed to build the ports in Leopard.
- We can now build the final package (.pkg) without administrative privileges.
- Using pmdoc file format to keep the meta datas of packages with PackageMaker.
- Using CFLAGS inspired in XCode defaults.
- Added many test features.
- Python modules are compiled for the native Python in Leopard.

* New ports
cairo, erlang, pixman, xar.

* Ports updated

- apr 1.3.3
- apr-util 1.3.4
- autoconf 2.63
- automake 1.10.1
- bzr 1.9
- dialog 1.1-20080819
- dictd 1.1.0
- django 1.0
- elinks 0.11.5
- ffmpeg 20081219
- git 1.6.0.3
- glib 2.18.2
- gmp 4.2.4
- gnuplot 4.2.4
- guile 1.8.6
- httpd 2.2.11
- imap 2007d
- lame 3.98.2
- lftp 3.7.6
- libevent 1.4.8
- libsigc++ 2.3.3
- libsigsegv 2.6
- libtorrent 0.12.2
- lighttpd 1.4.20
- lua 5.1.4
- lzma 4.32.7
- m4 1.4.12
- memcachedb 1.1.0
- mercurial 1.0.2
- mpop 1.0.15
- msmtp 1.4.16
- mtr 0.75
- neon 0.28.3
- nmap 4.76
- pcre 7.8
- popt 1.14
- python 3.0
- rdiff-backup 1.2.2
- rsync 3.0.4
- rtorrent 0.8.2
- rudix 1.7.1
- shuffle 0.9
- sox 14.1.0
- sqlite 3.6.4
- subversion 1.5.4
- varnish 2.0.2

CHANGES FROM RUDIX 2008 to RUDIX 2008R2

* Build System

$(NAMEVERSION) and $(NAMEVERSIONREV) tags are now absolete.

Added a new rule to test a package after build. Test focus Universal
Build and tests provided by package's original sources.

* New packages/ports

cheetah, coretuils, db, django, elinks, feedparser, frotz, gnuplot, hoc,
imap-uw, libpng, libsigsegv, lzma, m4, make, memcachedb, mercurial, psyco,
popt, python-docs, rdiff-backup, tomcat, varnish, zsync.

cheetah.pkg (cheetah-2.0.1-2.dmg)
Python-powered template engine and code generator.

coreutils.pkg (coreutils-6.9-2.dmg)
GNU core utilities.

db.pkg (db-4.7.25-3.dmg)
Berkeley DB package, revision 4.

django.pkg (django-0.96.2-3.dmg)
High-level Python web framework.

elinks.pkg (elinks-0.11.4-0.dmg)
Full-featured text WWW browser.

feedparser.pkg (feedparser-4.1-2.dmg)
RSS and Atom feeds parser written in Python.

frotz.pkg (frotz-2.43-2.dmg)
Curses-based interpreter for Infocom-compatible games.

gnuplot.pkg (gnuplot-4.2.3-2.dmg)
Command-driven interactive function plotting program.

hoc.pkg (hoc-1.1-2.dmg)
From Kernighan and Pike: High Order Calculator.

imap.pkg (imap-2007b-2.dmg)
University of Washington IMAP4rev1/POP2/POP3 mail servers.

libpng.pkg (libpng-1.2.29-1.dmg)
Library for manipulating PNG images.

libsigsegv.pkg (libsigsegv-2.5-2intel.dmg)
Library for handling page faults in user mode.a

lzma.pkg (lzma-4.32.6-2.dmg)
LZMA utils - make usage of LZMA compression easy.

m4.pkg (m4-1.4.11-2.dmg)
GNU m4.

make.pkg (make-3.81-0.dmg)
GNU make.

memcachedb.pkg (memcachedb-1.0.3-2.dmg)
Distributed key-value storage system.

mercurial.pkg (mercurial-1.0.1-1.dmg)
Fast, lightweight source control management.

psyco.pkg (psyco-1.6-3.dmg)
Extension module to speed up the execution of Python code.

popt.pkg (popt-1.10.4-2.dmg)
getopt(3)-like library with a number of enhancements.

python-docs.pkg (python-docs-2.5.2-1.dmg)
Python 2.5 documentation.

rdiff-backup.pkg (rdiff-backup-1.1.6-0.dmg)
Convenient and transparent local/remote incremental mirror/backup.

apache-tomcat.pkg (apache-tomcat-6.0.18-0.dmg)
Java servlet 2.5 and Java server pages 2.1 server.

varnish.pkg (varnish-1.1.2-1.dmg)
High-performance HTTP accelerator.

zsync.pkg (zsync-0.5-3.dmg)
Client-side implementation of the rsync algorithm.

* Packages/Ports updates

Upgraded apr and apr-util to version 1.3.2.
apr-util now depends on sqlite.
Upgraded autoconf to version 2.62.
Upgraded bzr to version 1.5.
Upgraded cadaver to version 0.23.2.
Upgraded cflow to version 1.2.
Upgraded dialog to version 1.1-20080316.
Upgraded dtach to version 0.8.
Upgraded epic4 to version 2.10.
Upgraded ffmpeg.
Upgraded findutils to version 4.4.0.
Upgraded gettexto to version 0.17.
Upgraded git to version 1.5.6.2.
Upgraded gnupg to version 1.4.9.
grep now depends on pcre.
Upgraded guile to version 1.8.5.
Upgraded hexedit to version 1.2.10.
Upgraded httpd to version 2.2.9.
httpd enabled with mod_proxy.
Upgraded lame to version 3.98.
Upgraded lftp to version 3.7.3.
Upgraded libevent to version 1.4.5.
Upgraded lighttpd to version 1.4.19.
Upgraded lua to version 5.1.3.
Upgraded memcached to version 1.2.6.
Upgraded mpg123 to version 1.4.3.
Upgraded mpop to version 1.0.14.
Upgraded msmtp to version 1.4.13.
Upgraded mtr to version 0.73.
Upgraded nmap to version 4.68.
Upgraded p7zip to version 4.58.
Upgraded pcre to version 7.7.
Upgraded pkg-config to version 0.23.
Upgraded python to version 2.5.2.
Upgraded rsync to version 3.0.3.
Upgraded ruby to version 1.8.7-p22.
Upgraded rudix to version 1.5.2.
Upgraded sdl to version 1.2.13.
Upgraded shuffle to version 0.6.
Upgraded sox to version 14.0.1.
Upgraded sqlite to version 3.6.0.
Upgraded subversion to version 1.5.0.
Upgraded tree to version 1.5.2.
Upgraded txt2tags to version 2.5.
Upgraded wget to version 1.11.4.
Upgraded xemacs to version 21.4.21.
Upgraded xmp to version 2.5.1.


CHANGES FROM RUDIX 2007R4 TO RUDIX 2008

Build System
------------

Rudix's packages are longer marked in ``/usr/local/lib/rudix/''.
Please, use rudix command line version 1.2 now.

Added Makefile variable ``LICENSE='' which allows one to select a pre-defined license.

File list in ReadMe.rtf are now automatically created in rule ``pkg''.

Added support to create pre/post install scripts, just create a folder ``scripts/'' and put the scripts there.

Universal binaries are so common now and they aren't longer marked as special in package's release name.

Intel only packages (or possible PowerPC too) are now marked as intel in package's release name.

Fixed a old typo, Makefile rule to retrieve was named retrive.

New Ports
---------

* dialog-1.1-20071028
- Display dialog boxes from shell scripts.


Ports/rudix/rudix.py
#!/usr/bin/env python

'''Rudix Package Manager -- RPM ;D

Usage:
rudix [help|version|available|aliases|list|remove-all|verify-all|update|interactive]
      [info <package-id>|files <package-id>|install <package-id>|remove <package-id>|search <package-id>|owner <path>|verify <package-id>|fix <package-id>]

rudix [-h|-v|-a|-A|-l|-R|-K|-u|-z]
      [-I <package-id>|-L <package-id>|-i <package-id>|-r <package-id>|-s <package-id>|-S <path>|-V <package-id>|-f <package-id>|-n <package-id>]

List all installed packages (package-id) unless options are given, like:
  -h    This help message
  -v    Print version
  -a    List all packages available for installation (name-version-release)
  -A    List all alternative aliases
  -l    List all installed packages (package-id, version and install date)
  -I    Print package information (package-id, version and install date)
  -L    List content of package (files)
  -i    Install package (download if not a file)
  -r    Remove package
  -R    Remove *all* Rudix packages installed (ask to confirm)
  -s    List available versions for package-id
  -S    Search for <path> in all packages and print if matched
  -V    Verify package
  -K    Verify all installed packages
  -f    Fix (repair) package
  -n    Download and install package (remote install)
  -u    Download and install all updated packages (remote update)
  -z    Interactive mode (type exit to quit)

Where <package-id> is either org.rudix.pkg.<name> or <name>.
'''

import sys
import os
import getopt
import tempfile
import re
from StringIO import StringIO
from gzip import GzipFile
from subprocess import Popen, PIPE, call
from urllib2 import urlopen, Request
from platform import mac_ver

__author__ = 'Ruda Moura'
__copyright__ = 'Copyright (c) 2005-2012 Ruda Moura'
__credits__ = 'Ruda Moura, Leonardo Santagada'
__license__ = 'BSD'
__version__ = '@VERSION@'

PROGRAM_NAME = os.path.basename(sys.argv[0])
PREFIX = 'org.rudix.pkg.'
OSX_VERSION = [int(x) for x in mac_ver()[0].split('.')[0:2]] # (MAJOR, MINOR)
RUDIX_NAMES = {
    (10, 6): 'rudix-snowleopard',
    (10, 7): 'rudix',
}
RUDIX = RUDIX_NAMES.get(tuple(OSX_VERSION), 'rudix')
VERSION = 2012

NAME_OPTS = {
    'help': '-h',
    'version': '-v',
    'aliases': '-A',
    'available': '-a',
    'list': '-l', 'ls': '-l', 'dir': '-l',
    'info': '-I', 'about': '-I',
    'install': '-i',
    'uninstall': '-r', 'remove': '-r',
    'uninstall-all': '-R', 'remove-all': '-R',
    'files': '-L', 'content': '-L',
    'search': '-s', 'versions': '-s',
    'owner': '-S',
    'verify': '-V', 'check': '-V',
    'verify-all': '-K',
    'fix': '-f',
    'update': '-u', 'upgrade': '-u',
    'repl': '-z', 'interactive': '-z',
}

ALIASES = {
    'aria': 'aria2',
    'awk': 'gawk',
    'bazaar': 'bzr',
    'exuberant': 'ctags', 'exuberant-ctags': 'ctags',
    'fab': 'fabric',
    'gnumake': 'make',
    'hg': 'mercurial',
    'memcache': 'memcached',
    'pkgconfig': 'pkg-config',
    'rdiff': 'librsync', 'rdiffbackup': 'rdiff-backup',
    'ssh': 'python-ssh',
    'svn': 'subversion',
    'supervisord': 'supervisor',
    'tofrodos': 'dos2unix',
    'tomcat': 'tomcat6',
    'unix2dos': 'dos2unix',
}

def rudix_version():
    'Print current Rudix version'
    print 'Rudix Package Manager version %s' % __version__
    print __copyright__

def usage():
    'Print help'
    print __doc__

def _is_root():
    'Test for root privileges'
    return os.getuid() == 0

def _root_required():
    'Requires root message'
    print >> sys.stderr, '%s: this operation requires root privileges' % PROGRAM_NAME
    return 1

def _communicate(args):
    'Call a process and return its stdout data as a list of strings'
    proc = Popen(args, stdout=PIPE, stderr=PIPE)
    return proc.communicate()[0].split('\n')[:-1]

def is_package_installed(pkg):
    'Test if package is installed'
    pkg = normalize(pkg)
    out = _communicate(['pkgutil', '--pkg-info', pkg])
    for line in out:
        if line.startswith('install-time: '):
            return True
    return False

def is_package_with_version_installed(pkg, version):
    'Test if package with version is installed'
    pkg = normalize(pkg)
    current = None
    out = _communicate(['pkgutil', '--pkg-info', pkg])
    for line in out:
        if line.startswith('version: '):
            current = line[len('version: '):]
        if line.startswith('install-time: '):
            break
    if current == None:
        return False
    if version_compare(current, version) == 0:
        return True
    else:
        return False

def get_packages():
    'Get a list of packages installed'
    out = _communicate(['pkgutil', '--pkgs=' + PREFIX + '*'])
    pkgs = [line.strip() for line in out]
    return pkgs

def get_package_info(pkg):
    'Get information from package'
    pkg = normalize(pkg)
    out = _communicate(['pkgutil', '-v', '--pkg-info', pkg])
    version = None
    install_date = None
    for line in out:
        line = line.strip()
        if line.startswith('version: '):
            version = line[len('version: '):]
        elif line.startswith('install-time: '):
            install_date = line[len('install-time: '):]
    return version, install_date

def get_package_content(pkg, filter_dirs=True):
    'Get a list of file names from package'
    pkg = normalize(pkg)
    out = _communicate(['pkgutil', '--files', pkg])
    content = ['/'+line.strip() for line in out]
    if filter_dirs:
        content = [x for x in content if os.path.isfile(x)]
    return content

def print_package_info(pkg):
    'Print information about pkg'
    version, install_date = get_package_info(pkg)
    if install_date is not None:
        print '%s version %s (install: %s)'%(pkg, version, install_date)
    else:
        print "No receipt for '%s' found at '/'."%pkg # pretend we are pkgutil

def list_all_packages():
    'List all packages installed'
    for pkg in get_packages():
        print pkg

def list_all_packages_info():
    'List all packages installed, more detailed'
    for pkg in get_packages():
        print_package_info(pkg)

def list_package_files(pkg):
    'Print the file names of package'
    content = get_package_content(pkg)
    for file in content:
        print file

def install_package(pkg):
    'Install a local package'
    if _is_root():
        call(['installer', '-pkg', pkg, '-target', '/'], stderr=PIPE)
    else:
        _root_required()

def remove_package(pkg):
    'Uninstall a package'
    if _is_root() == False:
        return _root_required()
    pkg = normalize(pkg)
    for x in get_package_content(pkg):
        try:
            os.unlink(x)
        except OSError, e:
            print >> sys.stderr, e
    with open('/dev/null') as devnull:
        call(['pkgutil', '--forget', pkg], stderr=devnull)

def remove_all_packages():
    'Uninstall all packages'
    if _is_root == False:
        return _root_required()
    print "Using this option will remove *all* Rudix's packages!"
    print "Are you sure you want to proceed? (answer 'yes' or 'y' to confirm)"
    answer = raw_input().strip()
    if answer not in ['yes', 'y']:
        print 'Good!'
        return
    print 'Removing packages...'
    for pkg in get_packages():
        print pkg
        remove_package(pkg)
    # remember LinuxConf...
    print 'Cry a little tear because Rudix was removed'

def search_in_packages(path):
    'Search for path in all packages'
    out = _communicate(['pkgutil', '--file-info', path])
    for line in out:
        line = line.strip()
        if line.startswith('pkgid: '):
            print line[len('pkgid: '):]

def verify_package(pkg):
    'Verify package sanity'
    pkg = normalize(pkg)
    call(['pkgutil', '--verify', pkg], stderr=PIPE)

def verify_all_packages():
    'Verify all packages sanity'
    for pkg in get_packages():
        verify_package(pkg)

def fix_package(pkg):
    'Try to fix permissions and groups of package'
    if _is_root() == False:
        return _root_required()
    pkg = normalize(pkg)
    call(['pkgutil', '--repair', pkg], stderr=PIPE)

def version_compare(v1, v2):
    'Compare software version'
    from distutils.version import LooseVersion
    # remove the release
    ver_rel_re = re.compile('([^-]+)(?:-(\d+)$)?')
    v1, r1 = ver_rel_re.match(v1).groups()
    v2, r2 = ver_rel_re.match(v2).groups()

    v_cmp = cmp(LooseVersion(v1), LooseVersion(v2))
    if v_cmp == 0:
        # if the same version then compare the release
        if r1 is None:
            r1 = 0
        if r2 is None:
            r2 = 0
        return cmp(int(r1), int(r2))
    else:
        return v_cmp


def _retrieve(url):
    'Retrieve content from URL'
    UA = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_7_3) AppleWebKit/534.53.11 (KHTML, like Gecko) Version/5.1.3 Safari/534.53.10'
    request = Request(url)
    request.add_header('Accept-Encoding', 'gzip')
    request.add_header('User-Agent', UA)
    response = urlopen(request)
    if response.headers.get('content-encoding', '') == 'gzip':
        buf = StringIO(response.read())
        gz = GzipFile(fileobj=buf)
        content = gz.read()
    else:
        content = response.read()
    response.close()
    return content

def _retrieve_simple(url):
    'Retrieve content from URL'
    data = urlopen(url)
    content = data.read()
    return content

def get_available_packages(rudix_version=VERSION, limit=1000):
    '''Get available packages.
    Return a list (ordered by release time, lastest first) of all packages available for installation.
    Filters: rudix_version and limit.
    '''
    url = 'http://code.google.com/p/%s/downloads/list?q=Rudix:%d&num=%d&can=2' % (RUDIX, rudix_version, limit)
    content = _retrieve(url)
    packages = re.findall('%s.googlecode.com/files/(.*)(\.dmg|\.pkg)"' % RUDIX, content)
    return packages

def print_available_packages():
    'Print all packages available for installation'
    versions = get_available_packages()
    for version in versions:
        name = version[0]
        print name

def print_aliases():
    'Print all aliases'
    aliases = ALIASES.keys()
    aliases.sort()
    for alias in aliases:
        print "alias '%s' for package '%s'" % (alias, ALIASES[alias])

def get_versions_for_package(pkg, rudix_version=VERSION, limit=10):
    'Get a list of available versions for package'
    pkg = denormalize(pkg)
    url = 'http://code.google.com/p/%s/downloads/list?q=Filename:%s+Rudix:%d&num=%d&can=2' % (RUDIX, pkg, rudix_version, limit)
    content = _retrieve(url)
    urls = re.findall('(%s.googlecode.com/files/(%s-([\w.]+(?:-\d+)?(?:.i386)?)(\.dmg|\.pkg)))' % (RUDIX, pkg), content)
    versions = sorted(list(set(urls)),
                      cmp=lambda x, y: version_compare(x[2], y[2]))
    if len(versions) == 0:
        return []
    else:
        return versions

def get_latest_version_of_package(pkg):
    'Get the latest version of package'
    versions = get_versions_for_package(pkg)
    if versions:
        return versions[-1]
    else:
        return []

def print_versions_for_package(pkg):
    'Print all versions available for package'
    versions = get_versions_for_package(pkg)
    for version in versions:
        name = version[1]
        if name.endswith('.dmg'):
            name = name[:name.index('.dmg')]
        if name.endswith('.pkg'):
            name = name[:name.index('.pkg')]
        print name

def net_install_package(pkg, net_info):
    'Support function for net_install_command'
    net_url, net_filename, net_version, net_extension = net_info
    print 'Downloading', net_url
    tempf, file_path = tempfile.mkstemp(suffix=net_extension)
    try:
        call(['curl', '-f', '-o', file_path, '-C', '-', '-L', '-#', net_url])
        if net_extension == '.dmg':
            print 'Mounting downloaded image file', file_path
            out = _communicate(['hdiutil', 'attach', '-noautoopen', file_path])
            for l in out:
                if 'Apple_partition_scheme' in l:
                    disk_path = l.split()[0]
                if 'Apple_HFS' in l:
                    volume_path = l.split()[2]
            filepath = os.path.join(volume_path, denormalize(pkg) + '.pkg')
        if net_extension == '.pkg':
            filepath = file_path
        if os.path.exists(filepath):
            install_package(filepath)
        if net_extension == '.dmg':
            print 'Unmounting image', volume_path
            call(['hdiutil', 'detach', disk_path], stdout=PIPE, stderr=PIPE)
    finally:
        os.close(tempf)

def net_install_command(pkg):
    'Install package from the Internet if the package was not installed or is older than the internet version'
    net_info = get_latest_version_of_package(pkg)
    version, install_date = get_package_info(pkg)
    if net_info == []:
        print "Package '%s' not found online" % pkg
        return
    if version is not None and version_compare(version, net_info[2]) >= 0:
        print 'Latest version of package %s(%s) already installed' % (pkg, version)
        return
    if _is_root():
        net_install_package(pkg, net_info)
        print 'All done'
    else:
        _root_required()

def update_all_packages():
    'Try to update the current base of packages'
    to_update = []
    # take each package, go to the internet and see if there is a newer version
    for pkg in get_packages():
        net_info = get_latest_version_of_package(pkg)
        version, install_date = get_package_info(pkg)
        if net_info == [] or version_compare(version, net_info[2]) >= 0:
            continue
        print '{0:25} {1:10} will be updated to version {2}'.format(denormalize(pkg), version, net_info[2])
        to_update.append((pkg, net_info))
    if len(to_update) == 0:
        print 'All packages are up to date'
        return
    # if there is packages to update you need to be root
    if _is_root():
        for pkg, net_info in to_update:
            net_install_package(pkg, net_info)
        print 'All done'
    else:
        _root_required()

def normalize(pkg):
    'Transform package in full pkg-id (with PREFIX)'
    if ALIASES.has_key(pkg):
        pkg = ALIASES[pkg]
    if not pkg.startswith(PREFIX):
        pkg = PREFIX + pkg
    return pkg

def denormalize(pkg):
    'Transform package in name without PREFIX'
    if pkg.startswith(PREFIX):
        pkg = pkg[len(PREFIX):]
    return pkg

def repl():
    'The interactive mode (read-eval-print-loop)'
    rudix_version()
    while True:
        print ']',
        try:
            line = raw_input().strip()
        except KeyboardInterrupt:
            print
            return 0
        except EOFError:
            print
            return 0
        if not line:
            continue
        if line in ['quit', 'exit', 'end', 'bye', 'halt']:
            break
        argv = [None] + line.split() # fake ARGV
        main(argv)

def process(args):
    'Process arguments and execute some action'
    try:
        opts, args = getopt.getopt(args, "aAhI:lL:i:r:Rs:S:vV:Kf:n:uz")
    except getopt.error, msg:
        print >> sys.stderr, '%s: %s'%(PROGRAM_NAME, msg)
        print >> sys.stderr, '\t for help use -h or help'
        return 2
    # option processing
    for option, value in opts:
        if option == '-h':
            usage()
            return 0
        if option == '-v':
            rudix_version()
            return 0
        if option == '-a':
            print_available_packages()
        if option == '-A':
            print_aliases()
        if option == '-I':
            print_package_info(normalize(value))
        if option == '-l':
            list_all_packages_info()
        if option == '-L':
            list_package_files(normalize(value))
        if option == '-i':
            if os.path.isfile(value):
                install_package(value)
            else:
                net_install_command(normalize(value))
        if option == '-r':
            remove_package(normalize(value))
        if option == '-R':
            remove_all_packages()
        if option == '-s':
            print_versions_for_package(normalize(value))
        if option == '-S':
            search_in_packages(value)
        if option == '-V':
            verify_package(normalize(value))
        if option == '-K':
            verify_all_packages()
        if option == '-f':
            fix_package(normalize(value))
        if option == '-n':
            net_install_command(normalize(value))
        if option == '-u':
            update_all_packages()
        if option == '-z':
            repl()
    if args:
        try:
            opt = NAME_OPTS[args[0]]
        except KeyError, e:
            print >> sys.stderr, "%s: unknown command %s"%(PROGRAM_NAME, e)
            print >> sys.stderr, '\t for help use -h or help'
            return 2
        args[0] = opt
        return process(args)

def main(argv=None):
    if argv is None:
        argv = sys.argv
    if len(argv) == 1:
        list_all_packages()
        return 0
    return process(argv[1:])

if __name__ == "__main__":
    sys.exit(main())
